### PR TITLE
Clean Up The Module Dependency Graph Format A Bit

### PR DESCRIPTION
**Pull request by CodaFi**
- Squash Fingerprint data into the end of a node record as a trailing blob.
- Restore the immutable `fingerprint` member
- Document, clean up, and formalize more serialization infrastructure.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci test

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift [hunk 1]
```diff
@@ -267,13 +267,35 @@ extension ModuleDependencyGraph {
 // MARK: - Serialization
 
 extension ModuleDependencyGraph {
+  /// The leading signature of this file format.
   fileprivate static let signature = "DDEP"
+  /// The expected version number of the serialized dependency graph.
+  ///
+  /// - WARNING: You *must* increment the minor version number when making any
+  ///            changes to the underlying serialization format.
+  fileprivate static let version = Version(1, 0, 0)
```
Comment (CodaFi) [inline Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift:276]: N.B. This warning is meant to ward people off, but we don't actually have to care about stability until we ship.

**Review by davidungar — APPROVED** [review]
LGTM!

**Comment by CodaFi** [comment]
⛵

**Pull request merged**